packages/runtime/src/next.tsx

```diff
@@ -158,3 +169,49 @@ export function Page({ page, rootElement, makeswiftApiEndpoint, cacheData, previ
     </RuntimeProvider>
   )
 }
+
+const FORWARDED_NEXT_DYNAMIC_REF_KEY = '__forwarded_next_dynamic_ref__'
```

Comment:
Does this code preserve forwarded refs through the dynamic element of the components?

Reply:
Yeah. Basically, `next/dynamic` uses `useImperativeHandle` and exposes a ref that looks something like this:

```ts
useImperativeHandle(ref, () => ({
  refetch() {
    // refetch the JS chunk
  }
}), [])
```

But we need that ref to give us the DOM node (or have no ref at all so we can automatically find it). So instead we kind of drill it _through_ the dynamic component by using this special key (`__forwarded_next_dynamic_ref__`) and then doing `ref={props['__forwarded_next_dynamic_ref__']}` in the `patch` function.